core/modules/wcontrol/Foreman.cc

```diff
@@ -38,65 +38,61 @@
 
 // Qserv headers
 #include "mysql/MySqlConfig.h"
+#include "proto/worker.pb.h"
 #include "wbase/Base.h"
+#include "wbase/SendChannel.h"
 #include "wconfig/Config.h"
-#include "wcontrol/RunnerMgr.h"
 #include "wdb/ChunkResource.h"
+#include "wdb/QueryRunner.h"
 
 
 namespace lsst {
 namespace qserv {
 namespace wcontrol {
 
-Foreman::Ptr Foreman::newForeman(Scheduler::Ptr const& sched) {
-    return std::make_shared<Foreman>(sched);
+Foreman::Ptr Foreman::newForeman(Scheduler::Ptr const& sched, uint poolSize) {
+    return std::make_shared<Foreman>(sched, poolSize);
 }
 
-Foreman::Foreman(Scheduler::Ptr const& s) : _scheduler{s} {
+Foreman::Foreman(Scheduler::Ptr const& s, uint poolSize) : _scheduler{s} {
     // Make the chunk resource mgr
     mysql::MySqlConfig c(wconfig::getConfig().getSqlConfig());
     _chunkResourceMgr = wdb::ChunkResourceMgr::newMgr(c);
-    _rManager.reset(new RunnerMgr(*this));
     assert(s); // Cannot operate without scheduler.
+
+    LOGF_DEBUG("poolSize=%1%" % poolSize);
+    _pool = util::ThreadPool::newThreadPool(poolSize, _scheduler);
 }
 Foreman::~Foreman() {
     LOGF(_log, LOG_LVL_DEBUG, "Foreman::~Foreman()");
     // FIXME: Cancel and drain runners. It will take significant effort to have xrootd shutdown cleanly
     // and this will never get called until that happens, making this a very low priority item.
     // This should only (get called on shutdown/restart).
+    _pool->endAll();
 }
 
 /// Create and queue a Task from a TaskMsg and a replyChannel.
 wbase::Task::Ptr Foreman::processMsg(std::shared_ptr<proto::TaskMsg> const& taskMsg,
                                      std::shared_ptr<wbase::SendChannel> const& replyChannel) {
     auto task = std::make_shared<wbase::Task>(taskMsg, replyChannel);
-    newTaskAction(task);
-    return task;
-}
-
-void Foreman::_startRunner(wbase::Task::Ptr const& t) {
-    auto f = [this](wbase::Task::Ptr t) {
-        LOGF_DEBUG("Foreman::_startRunner start");
-        RunnerMgr *rm = this->_rManager.get();
-        Runner::Ptr rp = std::make_shared<Runner>(rm, t);
-        (*rp)();
-        LOGF_DEBUG("Foreman::_startRunner end");
+    auto func = [this, task](){
+        proto::TaskMsg const& msg = *task->msg;
+        if(!msg.has_protocol() || msg.protocol() < 2) {
```

Comment:
I would use a descriptive const int or enum here. I don't know what protocol you are referring to other than a magic number.